ARCHITECTURE.md

```diff
@@ -0,0 +1,127 @@
+# Architecture
+
+This document describes high-level architecture of rust-analyzer.
+If you want to familiarize yourself with the code base, you are just
+in the right place!
+
+
+## Code generation
+
+Some of the components of this repository are generated through automatic
+processes. These are outlined below:
+
+- `gen-kinds`: The kinds of tokens are reused in several places, so a generator
+  is used. This process uses [tera] to generate, using data in [grammar.ron],
+  the files:
```

Comment:
I'd rewrite this sentence as: *We use tera templates to generate the files listed below, based on the grammar described in grammar.ron*